File: src/tools/run-command.ts
import { spawn } from "node:child_process";
import { performance } from "node:perf_hooks";

import type { ToolDefinition } from "../core/types.js";
import { HeadTailBuffer } from "../text/head-tail-buffer.js";
import { safeDisplay } from "../text/safe-display.js";
import type { Workspace } from "../workspace/workspace.js";
import {
  objectSchema,
  optionalInteger,
  parseArgumentObject,
  requiredString,
} from "./arguments.js";
import {
  type LocalTool,
  type PreparedToolCall,
  failure,
  success,
} from "./types.js";

const DEFAULT_TIMEOUT_SECONDS = 60;
const MAX_TIMEOUT_SECONDS = 120;
const TERMINATION_GRACE_MS = 500;

export class RunCommandTool implements LocalTool {
  public readonly effect = "execute" as const;
  public readonly definition: ToolDefinition = {
    name: "run_command",
    description:
      "Run a shell command in the workspace. Returns bounded combined output, exit status, and duration. This is not an OS sandbox.",
    parameters: objectSchema(
      {
        command: { type: "string", description: "Shell command to run" },
        timeout_seconds: {
          type: "integer",
          minimum: 1,
          maximum: MAX_TIMEOUT_SECONDS,
          description: `Timeout in seconds, default ${DEFAULT_TIMEOUT_SECONDS}`,
        },
      },
      ["command"],
    ),
  };

  readonly #workspace: Workspace;

  public constructor(workspace: Workspace) {
    this.#workspace = workspace;
  }

  public prepare(argumentsJson: string): PreparedToolCall {
    const object = parseArgumentObject(argumentsJson, ["command", "timeout_seconds"]);
    const command = requiredString(object, "command", { nonBlank: true });
    const timeoutSeconds = optionalInteger(
      object,
      "timeout_seconds",
      DEFAULT_TIMEOUT_SECONDS,
      1,
      MAX_TIMEOUT_SECONDS,
    );
    const compact = command.trim().replace(/\s+/g, " ");
    const preview = safeDisplay(compact);
    return {
      preview,
      execute: async (signal) =>
        await runCommand(command, timeoutSeconds, this.#workspace.root, signal),
    };
  }
}

async function runCommand(
  commandText: string,
  timeoutSeconds: number,
  workingDirectory: string,
  signal: AbortSignal,
) {
  if (signal.aborted) {
    return failure("command was cancelled before it started", "cancelled");
  }

  const output = new HeadTailBuffer(48 * 1024, 16 * 1024);
  const started = performance.now();
  const detached = process.platform !== "win32";

  return await new Promise<ReturnType<typeof success>>((resolve) => {
    const child = spawn("/bin/sh", ["-lc", commandText], {
      cwd: workingDirectory,
      env: sanitizedEnvironment(process.env),
      detached,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let timedOut = false;
    let cancelled = false;
    let settled = false;
    let forceKillTimer: NodeJS.Timeout | undefined;

    child.stdout.on("data", (chunk: Buffer) => output.write(chunk));
    child.stderr.on("data", (chunk: Buffer) => output.write(chunk));

    const terminate = (): void => {
      if (child.pid === undefined || child.killed) {
        return;
      }
      killProcess(child.pid, "SIGTERM", detached);
      forceKillTimer = setTimeout(() => {
        if (child.pid !== undefined) {
          killProcess(child.pid, "SIGKILL", detached);
        }
      }, TERMINATION_GRACE_MS);
      forceKillTimer.unref();
    };

    const timeout = setTimeout(() => {
      timedOut = true;
      terminate();
    }, timeoutSeconds * 1_000);
    timeout.unref();

    const onAbort = (): void => {
      cancelled = true;
      terminate();
    };
    signal.addEventListener("abort", onAbort, { once: true });

    const finish = (
      exitCode: number,
      exitSignal: NodeJS.Signals | null,
      spawnError?: Error,
    ): void => {
      if (settled) {
        return;
      }
      settled = true;
      clearTimeout(timeout);
      if (forceKillTimer !== undefined) {
        clearTimeout(forceKillTimer);
      }
      signal.removeEventListener("abort", onAbort);
      const durationMs = Math.max(0, Math.round(performance.now() - started));
      const rawOutput = output.toString();
      const retainedOutput = rawOutput.trim() === "" ? "(command produced no output)" : rawOutput;
      const summary = `exit=${exitCode} duration=${durationMs}ms\n${retainedOutput}`;
      const metadata = {
        exit_code: exitCode,
        signal: exitSignal,
        duration_ms: durationMs,
        output_bytes: output.totalBytes,
        truncated: output.truncated,
      };

      if (spawnError !== undefined) {
        resolve(failure(`cannot start command: ${spawnError.message}`, "spawn_error", metadata));
      } else if (cancelled) {
        resolve(failure(summary, "cancelled", metadata));
      } else if (timedOut) {
        resolve(failure(summary, "timeout", metadata));
      } else if (exitCode !== 0) {
        resolve(failure(summary, "command_failed", metadata));
      } else {
        resolve(success(summary, metadata));
      }
    };

    child.once("error", (error) => finish(-1, null, error));
    child.once("close", (code, exitSignal) => finish(code ?? -1, exitSignal));
  });
}

function killProcess(pid: number, signal: NodeJS.Signals, processGroup: boolean): void {
  try {
    process.kill(processGroup ? -pid : pid, signal);
  } catch (error) {
    const code = (error as NodeJS.ErrnoException).code;
    if (code !== "ESRCH") {
      try {
        process.kill(pid, signal);
      } catch {
        // The process may have exited between the two kill attempts.
      }
    }
  }
}

export function sanitizedEnvironment(
  environment: NodeJS.ProcessEnv,
): NodeJS.ProcessEnv {
  const result: NodeJS.ProcessEnv = {};
  for (const [key, value] of Object.entries(environment)) {
    if (value !== undefined && !isSensitiveEnvironmentKey(key)) {
      result[key] = value;
    }
  }
  result.NASH_AGENT = "1";
  return result;
}

function isSensitiveEnvironmentKey(key: string): boolean {
  const upper = key.toUpperCase();
  if (
    upper === "SSH_AUTH_SOCK" ||
    upper === "GIT_ASKPASS" ||
    upper === "DEEPSEEK_API_KEY" ||
    upper === "NASH_API_KEY"
  ) {
    return true;
  }
  return [
    "API_KEY",
    "ACCESS_KEY",
    "PRIVATE_KEY",
    "TOKEN",
    "SECRET",
    "PASSWORD",
    "PASSWD",
    "CREDENTIAL",
    "COOKIE",
    "BEARER",
  ].some((fragment) => upper.includes(fragment));
}

File: internal/tool/run_command.go
package tool

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/TmkShink/nash-agent/internal/core"
	"github.com/TmkShink/nash-agent/internal/textlimit"
	"github.com/TmkShink/nash-agent/internal/workspace"
)

const (
	defaultCommandTimeout = 60 * time.Second
	maxCommandTimeout     = 120 * time.Second
)

type RunCommand struct {
	workspace *workspace.Workspace
}

func NewRunCommand(workspace *workspace.Workspace) *RunCommand {
	return &RunCommand{workspace: workspace}
}

func (t *RunCommand) Definition() core.ToolDefinition {
	return core.ToolDefinition{
		Name:        "run_command",
		Description: "Run a shell command in the workspace and return combined stdout/stderr, exit status, and duration. Execution is time- and output-bounded but not an OS sandbox.",
		InputSchema: objectSchema(map[string]any{
			"command":         map[string]any{"type": "string", "description": "Shell command to run"},
			"timeout_seconds": map[string]any{"type": "integer", "minimum": 1, "maximum": 120, "description": "Timeout in seconds, default 60"},
		}, "command"),
	}
}

func (*RunCommand) Effect() Effect { return EffectExecute }

func (*RunCommand) Preview(arguments string) string {
	var args struct {
		Command string `json:"command"`
	}
	if err := DecodeArguments(arguments, &args); err != nil {
		return "run command"
	}
	const maxPreview = 240
	preview := strings.TrimSpace(args.Command)
	if len(preview) > maxPreview {
		preview = preview[:maxPreview] + "..."
	}
	return preview
}

func (t *RunCommand) Execute(ctx context.Context, arguments string) Result {
	var args struct {
		Command        string `json:"command"`
		TimeoutSeconds int    `json:"timeout_seconds"`
	}
	if err := DecodeArguments(arguments, &args); err != nil {
		return Failure("invalid run_command arguments: "+err.Error(), map[string]any{"kind": "invalid_arguments"})
	}
	if strings.TrimSpace(args.Command) == "" {
		return Failure("command is required", map[string]any{"kind": "invalid_arguments"})
	}
	timeout := defaultCommandTimeout
	if args.TimeoutSeconds != 0 {
		timeout = time.Duration(args.TimeoutSeconds) * time.Second
	}
	if timeout <= 0 || timeout > maxCommandTimeout {
		return Failure("timeout_seconds must be between 1 and 120", map[string]any{"kind": "invalid_arguments"})
	}

	commandCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	command := exec.CommandContext(commandCtx, "/bin/sh", "-lc", args.Command)
	command.Dir = t.workspace.Root()
	command.Env = sanitizedEnvironment(os.Environ())
	// A killed shell can leave descendants holding stdout or stderr open. Bound
	// that pipe-drain wait even though process-tree isolation is out of scope.
	command.WaitDelay = 2 * time.Second
	output := textlimit.NewHeadTailBuffer(48*1024, 16*1024)
	command.Stdout = output
	command.Stderr = output

	started := time.Now()
	err := command.Run()
	duration := time.Since(started)
	exitCode := 0
	kind := ""
	if err != nil {
		exitCode = -1
		kind = "command_failed"
		var exitError *exec.ExitError
		if errors.As(err, &exitError) {
			exitCode = exitError.ExitCode()
		}
		if errors.Is(commandCtx.Err(), context.DeadlineExceeded) {
			kind = "timeout"
		}
	}

	content := output.String()
	if strings.TrimSpace(content) == "" {
		content = "(command produced no output)"
	}
	summary := fmt.Sprintf("exit=%d duration=%s\n%s", exitCode, duration.Round(time.Millisecond), content)
	metadata := map[string]any{
		"exit_code":    exitCode,
		"duration_ms":  duration.Milliseconds(),
		"output_bytes": output.TotalBytes(),
		"truncated":    output.Truncated(),
	}
	if err != nil {
		metadata["kind"] = kind
		return Failure(summary, metadata)
	}
	return Success(summary, metadata)
}

func sanitizedEnvironment(environment []string) []string {
	result := make([]string, 0, len(environment)+1)
	for _, entry := range environment {
		key, _, found := strings.Cut(entry, "=")
		if !found || sensitiveEnvironmentKey(key) {
			continue
		}
		result = append(result, entry)
	}
	return append(result, "NASH_AGENT=1")
}

func sensitiveEnvironmentKey(key string) bool {
	upper := strings.ToUpper(key)
	if upper == "SSH_AUTH_SOCK" || upper == "GIT_ASKPASS" || upper == "NASH_API_KEY" {
		return true
	}
	for _, fragment := range []string{"API_KEY", "TOKEN", "SECRET", "PASSWORD", "CREDENTIAL"} {
		if strings.Contains(upper, fragment) {
			return true
		}
	}
	return false
}
